File: web-user/src/api/balance.ts
import { api } from './client'

export interface LedgerItem {
  id: number
  delta_seconds: number
  reason: string
  ref_type: string | null
  ref_id: number | null
  balance_after: number
  note: string | null
  created_at: string
}

export interface LedgerPage {
  items: LedgerItem[]
  total: number
  page: number
  size: number
}

export async function listLedger(page = 1, size = 20): Promise<LedgerPage> {
  const res = await api.get<LedgerPage>('/api/balance/ledger', { params: { page, size } })
  return res.data
}

File: backend/app/schemas/balance.py
"""DTOs for balance ledger."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class LedgerItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    delta_seconds: int
    reason: str
    ref_type: Optional[str] = None
    ref_id: Optional[int] = None
    balance_after: int
    note: Optional[str] = None
    created_at: datetime


class LedgerPageOut(BaseModel):
    items: list[LedgerItemOut]
    total: int
    page: int
    size: int
